bookshelf/main_test.py

```diff
@@ -0,0 +1,183 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import re
+
+import google.auth
+import main
+import pytest
+import requests
+from retrying import retry
+from six import BytesIO
+
+
+credentials, project_id = google.auth.default()
+os.environ['GOOGLE_CLOUD_PROJECT'] = project_id
+
+
+@pytest.yield_fixture
+def app(request):
+    """This fixtures provides a Flask app instance configured for testing.
+
+    It also ensures the tests run within a request context, allowing
+    any calls to flask.request, flask.current_app, etc. to work."""
+    app = main.app
+
+    with app.test_request_context():
+        yield app
+
+
+@pytest.yield_fixture
+def firestore():
+
+    import firestore
+    """This fixture provides a modified version of the app's firebase model that
+    tracks all created items and deletes them at the end of the test.
+
+    Any tests that directly or indirectly interact with the database should use
+    this to ensure that resources are properly cleaned up.
+    """
+
+    # Ensure no books exist before running. This typically helps if tests
+    # somehow left the database in a bad state.
+    delete_all_books(firestore)
+
+    yield firestore
+
+    # Delete all books that we created during tests.
+    delete_all_books(firestore)
+
+
+# The backend data stores can sometimes be flaky. It's useful to retry this
+# a few times before giving up.
+@retry(
+    stop_max_attempt_number=3,
+    wait_exponential_multiplier=100,
+    wait_exponential_max=2000)
+def delete_all_books(firestore):
+    while True:
+        books, _ = firestore.next_page(limit=50)
+        if not books:
+            break
+        for book in books:
+            firestore.delete(book['id'])
+
+
+# Mark all test cases in this class as flaky, so that if errors occur they
+# can be retried. This is useful when databases are temporarily unavailable.
+# @flaky(rerun_filter=flaky_filter)
+# Tell pytest to use both the app and model fixtures for all test cases.
+# This ensures that configuration is properly applied and that all database
+# resources created during tests are cleaned up. These fixtures are defined
+# in conftest.py
+def test_list(app, firestore):
+    for i in range(1, 12):
+        firestore.create({'title': u'Book {0}'.format(i)})
+
+    with app.test_client() as c:
+        rv = c.get('/')
+
+    assert rv.status == '200 OK'
+
+    body = rv.data.decode('utf-8')
+    assert 'Book 1' in body, "Should show books"
+    assert len(re.findall('<h4>Book', body)) <= 10, (
+        "Should not show more than 10 books")
+    assert 'More' in body, "Should have more than one page"
+
+
+def test_add(app):
+    data = {
+        'title': 'Test Book',
+        'author': 'Test Author',
+        'publishedDate': 'Test Date Published',
```

Comment:
Nit: should we use a date-type value for `publishedDate`?

Reply:
I looked into it, and it adds a surprising amount of complexity:

https://github.com/GoogleCloudPlatform/getting-started-python/commit/4ebbf58682696397b152c69775173b641804706f

My preference is to keep it simple and leave it as-is

Reply:
Got it - makes sense.